Update README.md figure

carolina-museum opened the pull request:
#63 
Changed the RAG architecture figure into a more appropriate one.

huyfififi commented on README.md:
```diff
@@ -41,9 +41,9 @@ RAG is a method that tries to avoid hallucination by using information that is p
 
 <p align="center">
-<img width="1225" alt="Screenshot 2024-10-12 at 02 27 44" src="https://github.com/user-attachments/assets/bf6537c6-d50e-45cc-bb2f-caac0f3d90fe">
+	<img width="1004" alt="Screenshot 2024-12-02 at 03 36 32" src="https://github.com/user-attachments/assets/0a752600-ed20-45eb-a0db-d8539dcb2465">
```
`nit` I'm not confident, but "embedded" and "embedding" are completely different things, and using embedded might be misleading...? 🤔

huyfififi replied:
is it common to use LLM twice? one is to combine knowledge and user query, and the other is to actually generate an answer

huyfififi replied:
I'm glad you are becoming familiar with diagrams.net 👍

carolina-museum replied:
I will check how embedded and embedding are different and make it clearer....!

I don't know how common it is, but using LLM twice does not seem necessary. I will see how common it is. We are not using it twice in our project right now either. 🤔 🤔 

Thank you for instructing me to use it! That helps my life a lot :)

carolina-museum replied:
I'll change the figure too if using LLM twice is not common!

carolina-museum replied:
I don't think using LLM twice is common, I think my understanding when I made the previous figure was incorrect.
I looked up some posts about RAG, and most of them didn't use LLM twice. 
This page explains several kinds of RAG architecture: https://humanloop.com/blog/rag-architectures

I will change the figure👍

carolina-museum replied:
I don't think embedded and embedding have different meanings, but what do you think?

I feel the usage is like this:
I embed the sequence every day as my day job. 
I embedded the sequence last night.
The sequence is embedded by an expert.
I am embedding the sequence right now. 
Here is an embedding of the sequence.

huyfififi replied:
@carolina-museum 
I think you are right, but I have never encountered "embed" being used in the context of generating embeddings though it is grammatically correct to use "embed" for this case (I don't know why 🤔 ) Instead, "embedding" is more commonly used as a noun, for example, when referring to "storing word embeddings."

I also browsed some web pages and searched for "embed(ed)", but I could not find it being used in this way on these pages.
- https://www.cloudflare.com/learning/ai/what-are-embeddings/
- https://en.wikipedia.org/wiki/Word_embedding
- https://en.wikipedia.org/wiki/Embedding
- https://aws.amazon.com/what-is/embeddings-in-machine-learning/
- https://pytorch.org/docs/stable/generated/torch.nn.Embedding.html
- https://learn.microsoft.com/en-us/azure/architecture/ai-ml/guide/rag/rag-generating-embeddings

but maybe I'm saying something stupid. Because I did not come up with helpful resources about this topic, let me just mark this thread as resolved

carolina-museum replied:
I see, thank you for the extended research!
I see that people try using the noun form most of the time. I added the noun usage too in the description.

For me, it sounds more intuitive to use "embedded" because it implies that the embeddings are made from what is being embedded. But! I will use more "embeddings" next time!

huyfififi commented on README.md:
```diff
@@ -41,9 +41,10 @@ RAG is a method that tries to avoid hallucination by using information that is p
 
 <p align="center">
-<img width="1225" alt="Screenshot 2024-10-12 at 02 27 44" src="https://github.com/user-attachments/assets/bf6537c6-d50e-45cc-bb2f-caac0f3d90fe">
+	<img width="800" alt="Screenshot 2024-12-02 at 19 48 11" src="https://github.com/user-attachments/assets/9c502a2c-234d-448e-aba9-8f961c6e5c1a">
```
I think the description of the generation part is outdated. That one should be responsible for generating answers instead of prompts

carolina-museum replied:
That's right! I will change it

huyfififi commented:
I'm getting confused... LLM is not used to generate prompts that will be used later in the stage, i.e., LLM for generation is only used once. Am I correct?

carolina-museum commented:
Yes, you are right!
I'll check my wording again

huyfififi commented:
@carolina-museum if you are still working on your changes, plz convert your pull request into "draft". Once you're done, make it a normal pull request and re-request review in the reviewer section so that I know when to leave my review

carolina-museum commented:
I finally got why I made you so confused!!!!
The generator in Lewis paper generated the next Token, but the generator in this README figure only concatenates sequences. I will upload a figure with this expressed.

carolina-museum commented:
I'm also thinking how I can update the other document (Lewis paper explanation) to make it more clear

huyfififi commented:
In the paper, they might suggest using an LLM twice, but I feel that the pattern of using an LLM for prompt generation isn't very common (though I'm not entirely sure; you might have a better perspective).
Therefore, I was thinking that removing the prompt generation part from the README (or simplify that part) would better reflect our system and make it easier for beginners to understand RAG by simplifying the explanation.

carolina-museum commented:
I got it about making it a draft!

carolina-museum commented:
That is a great idea to remove that part. That would be so much easier to understand!

huyfififi commented:
@carolina-museum Can you try rebasing (or merging) with the latest main branch? It's always a good idea to catch up with the target branch to avoid weird conflicts

huyfififi commented on README.md:
```diff
@@ -34,16 +34,16 @@ RAG is a method that tries to avoid hallucination by using information that is p
+1. A **user asks a question** (inputs a query).
+2. **k most related contexts to the query are retrieved** from the database. The distance between the embedded query and the embedded context determines the similarity between the query and the contexts.
+3. The k most related contexts(in text) are combined with the user's query(in text) to make a better prompt. It may simply be a **concatenation of the query and contexts with additional prompts**. Some works add prompts to respond "I don't know" if the information is not present in the context to minimize hallucinations.
+5. The better prompt made in the last step is fed to the LLM model, and the **LLM model gives the user an answer** based on the better prompt. The user receives this answer.
```
excuse me for nitpicking too much today 😂 but I think this number should be "4" instead of "5", though it shows the number correctly on UI

carolina-museum replied:
ahhhhhhh thank youuuuuuu
and no problem! I love your neat and precise review!

carolina-museum commented:
I see, let me try! 
Thank you for the advice!

carolina-museum commented:
I think it is rebased now! 🙏

carolina-museum commented:
maybe not... I'll try again

carolina-museum commented:
Is this good now...? It only seems like I added one new file. It is the exact difference, but this looks different from rebasing.

huyfififi commented:
![Screenshot 2024-12-03 at 2 42 44 AM](https://github.com/user-attachments/assets/015276ed-f2ed-4d33-b3ab-0116594c8d31)

something happening... the branch is behind 2 commits from the latest main 🤔 could I know what you did?
BTW, struggling with Git will be beneficial for you in the long run

carolina-museum commented:
Hmmm I'll think over it.
Does this indicate anything? ↓

<img width="470" alt="Screenshot 2024-12-03 at 03 02 10" src="https://github.com/user-attachments/assets/fd66a061-f23d-4c51-bebb-b603783ffb1f">

carolina-museum commented:
And yes, I think playing around with git like this helps me a lot!

huyfififi commented:
The `git status` does not offer much information... but how about you run `history` command so that I may be able to take a look at the whole command sequence?
https://www.geeksforgeeks.org/history-command-in-linux-with-examples/

huyfififi commented:
and `git log --graph` might provide you with the commit histories of `main` and your feature branch

carolina-museum commented:
I tried so many things so it might not be as helpful, but here is the history:
<img width="350" alt="Screenshot 2024-12-03 at 03 08 14" src="https://github.com/user-attachments/assets/0381c606-bed4-4b69-be64-7587da075d2f">

Thank you for taking a look!

carolina-museum commented:
and here is the graph. I'm taking a look at those right now.
I don't think I've done what I wanted to do😵‍💫

<img width="679" alt="Screenshot 2024-12-03 at 03 10 18" src="https://github.com/user-attachments/assets/2a35d8fe-82a4-4af2-a673-7769d7cc5474">

huyfififi commented:
is it possible to run `history 100`?

carolina-museum commented:
Here it is:

 1020  git branch fix-very-beginners-document
 1021  git status
 1022  git switch main
 1023  git branch
 1024  git branch -D fix-very-beginners-document
 1025  git pull
 1026  git branch fix-very-beginners-document
 1027  git switch fix-very-beginners-document
 1028  git status
 1029  git branch
 1030  git add run_project.md
 1031  git status
 1032  git commit -m "made changes based on review comments"
 1033  git push
 1034  git push --set-upstream origin fix-very-beginners-document
 1035  git branch
 1036  git switch main
 1037  git branch
 1038  git pull
 1039  git branch explain-p-lewis-2021-paper
 1040  git status
 1041  git add nlp_basics_articles/Paper_P_Lewis_2021.md
 1042  git commit -m "explaine P.Lewiis 2021 paper"
 1043  git push
 1044  git push --set-upstream origin explain-p-lewis-2021-paper
 1045  git branch
 1046  git pull
 1047  git branch
 1048  git checkout readme-figure-update
 1049  git status
 1050  git rebase main
 1051  git ls
 1052  ls
 1053  cd nlp_basics_articles
 1054  ls
 1055  git branch
 1056  git switch main
 1057  ls
 1058  git pull
 1059  ls
 1060  git checkout readme-figure-update
 1061  ls
 1062  git rebase main
 1063  ls
 1064  git add -u
 1065  git status
 1066  git commit -m "trying to rebase"
 1067  cd ../
 1068  git push
 1069  git pull
 1070  git config pull.rebase true
 1071  git add -u
 1072  git status
 1073  ls
 1074  cd nlp_basics_articles
 1075  ls
 1076  cd ../
 1077  cat README.md
 1078  git push
 1079  git log
 1080  git pull --rebase origin master
 1081  git pull --rebase origin main
 1082  git add -u
 1083  git status
 1084  git log
 1085  git log --graph
 1086  git branch
 1087  git push
 1088  git pull
 1089  git pull --reapply-cherry-picks
 1090  ls
 1091  git add nlp_basics_articles/Paper_P_Lewis_2021.md
 1092  git status
 1093  git push
 1094  git add --all
 1095  git status
 1096  git restore --staged chroma.sqlite3
 1097  git status
 1098  git status
 1099  history
 1100  git log -graph
 1101  git log --graph

huyfififi commented:
by the way, when doing copy & paste shell output, please use 

```
code
block
like
this
```

it can be done with Markdown

carolina-museum commented:
updated my comment :)
Thank you!

huyfififi commented:
Hmm... what will happen when you run `git rebase main` now?

carolina-museum commented:
Here...!

<img width="479" alt="Screenshot 2024-12-03 at 03 19 38" src="https://github.com/user-attachments/assets/c6a74ae1-c9cb-40a4-9921-bbc03cc98f52">

huyfififi commented:
can you show me the log graph now?

carolina-museum commented:
Here!
<img width="612" alt="Screenshot 2024-12-03 at 03 21 31" src="https://github.com/user-attachments/assets/598470d8-3d09-4f3c-a401-18a0a05e39a0">

Thank you so much🙏

carolina-museum commented:
```
carolina@Carolinas-MacBook-Air-2 rag % git log --graph
* commit c1fe8b703e61cbf856df5622d540e14766e63fec (HEAD -> readme-figure-update)
| Author: Carolina <carolinaminamioguchi@gmail.com>
| Date:   Tue Dec 3 01:57:01 2024 -0800
| 
|     5 --> 4 README.md
| 
* commit b39d152b1e087bf52443177c1d94c970554dc126
| Author: Carolina <carolinaminamioguchi@gmail.com>
| Date:   Tue Dec 3 01:46:47 2024 -0800
| 
|     eliminated the usage of generator in description
| 
* commit e9c23ee512dadd8db349079a8f7cf9bfa9a55eb5
| Author: Carolina <carolinaminamioguchi@gmail.com>
| Date:   Tue Dec 3 01:45:28 2024 -0800
| 
|     remove generator from RAG figure
| 
* commit c59ddd515cd17b1c0bc4f22b476e0dd22a5a4885
| Author: Carolina <carolinaminamioguchi@gmail.com>
| Date:   Tue Dec 3 01:10:26 2024 -0800
| 
|     RAG figure. use LLM only once
| 
* commit 849a710e3c1a0ffaea475edb90012d88972b2536
| Author: Carolina <carolinaminamioguchi@gmail.com>
| Date:   Tue Dec 3 00:06:32 2024 -0800
| 
|     changed my wording around generator
| 
* commit a6c0e1b7c952f68aa0ba3344e51eb13dae43a380
| Author: Carolina <carolinaminamioguchi@gmail.com>
| Date:   Mon Dec 2 23:29:35 2024 -0800
| 
|     forgot to bold one part in README.md :(
| 
* commit 0814bced24fe722240545a9f7b0b01c8c43ff898
| Author: Carolina <carolinaminamioguchi@gmail.com>
| Date:   Mon Dec 2 23:20:28 2024 -0800
| 
|     update RAG flow description and figure in README.md
|     
|     I updated the RAG flow explanation to be more accurate.
|     Also fixed the figure a bit to remove overlapped descriptions about k-most related contexts.
| 
* commit 14851b2f7ecbce137ee2a56e1888867b819fc5aa
| Author: Carolina <carolinaminamioguchi@gmail.com>
| Date:   Mon Dec 2 19:49:56 2024 -0800
| 
|     eliminate one LLM model from README.md figure
|     
|     Using two LLM models in one RAG structure is not common. I updated so that the figure shows the general form, using LLM just once.
| 
* commit 538c7fa97075ada277c82ec89971ad872bc9d0cf
| Author: Carolina <carolinaminamioguchi@gmail.com>
| Date:   Mon Dec 2 03:43:24 2024 -0800
| 
|     removed commented portion
| 
* commit a1a33b67064d2ffba53d0ffd642238cf63e3bf66
| Author: Carolina <carolinaminamioguchi@gmail.com>
| Date:   Mon Dec 2 03:42:21 2024 -0800
| 
|     Update README.md figure
|     
|     #63
|     Changed the RAG architecture figure into a more appropriate one.
| 
* commit c689d49362145d8366bdc71c5216ecc7371731f6 (origin/main, origin/HEAD, main)
| Author: Carolina <carolinaminamioguchi@gmail.com>
| Date:   Tue Dec 3 01:51:19 2024 -0800
| 
|     add meaning of freeze Paper_P_Lewis_2021.md (#90)
|     
|     * add meaning of freeze Paper_P_Lewis_2021.md
|     
|     I added more context on what freezing the learning means.🥶
|     
|     * Explain retrieval model more in Paper_P_Lewis_2021.md
|     
|     I added an explanation of why the retrieval model is considered a learnable model in this paper.
|     They use a retrieval model proposed in: https://arxiv.org/abs/1702.08734
|     This model is a neural model that we can train.
```

carolina-museum commented:
is `git push` right...?

<img width="612" alt="Screenshot 2024-12-03 at 03 26 14" src="https://github.com/user-attachments/assets/7243813b-114b-4d78-b7f1-b9d9de16f95f">

huyfififi commented:
Yes, and because we're overriding the history, you need to execute `git push -f` or `git push --force`

huyfififi commented:
now it's working 👍 you can navigate to the branch page and confirm that this branch is not behind the main branch

carolina-museum commented:
yayyyyyy thank you sooooo much!!!!

huyfififi commented:
do you still have some time today? There are many commits in the history though the changes are very small. I think it's better to combine the 10 commits into 1. it is totally fine if you call it a day. I'll go ahead and approve this pull request if so

carolina-museum commented:
I still have about 25 min until the place closes :)
I'll try combining the tiny commits into one!

carolina-museum commented:
Is it squash and merge? Or is it something else? I am still looking up right now

Thank you for all your instructions! I am making progress to enjoy my life more thanks to your support :)))

huyfififi commented:
1.

```
git rebase -i main
```

to initiate interactive rebase https://thoughtbot.com/blog/git-interactive-rebase-squash-amend-rewriting-history#interactive-rebase

2. replace `pick` with `s` or `squash` to combine the commit(s) into the previous one

https://thoughtbot.com/blog/git-interactive-rebase-squash-amend-rewriting-history#squash-commits-together

3. Follow the prompts
4. Because we changed the history, push with force

---

Let's try :-) if it's too hard to finish in 20 minutes, we can merge this

carolina-museum commented:
I think I got it!

huyfififi reviewed (approved):
Awesome!!! Now this pull request contains only one commit 😉 
![Screenshot 2024-12-03 at 3 47 55 AM](https://github.com/user-attachments/assets/1419e5a8-51f6-48f5-8a84-d88531bb580e)

It took much more time when I first tried to squash commits. You're doing great 👍

carolina-museum commented:
Yaaaaayyyyyyy
I struggled to get out of the interactive mode, but I made it! 
Your instructions were very clear and that helped me a lot!!

Thank you sooooooooo much!!
